packages/@eventual/core-runtime/src/handlers/entity-stream-worker.ts

```diff
@@ -45,20 +44,34 @@ export function createEntityStreamWorker(
 
         return { failedItemIds: result?.failedItemIds ?? [] };
       } else {
+        const itemsByKey: Record<string, EntityStreamItem<any>[]> = {};
+        items.forEach((item) => {
+          const normalizedKey = normalizeCompositeKey(entity, item.key);
+          const serializedKey = JSON.stringify(normalizedKey);
+          (itemsByKey[serializedKey] ??= []).push(item);
+        });
+
         const results = await promiseAllSettledPartitioned(
-          items,
-          async (item) => {
-            return await streamHandler.handler(item, context);
+          Object.entries(itemsByKey),
+          async ([, itemGroup]) => {
+            for (const i in itemGroup) {
+              const item = itemGroup[i]!;
+              try {
+                const result = await streamHandler.handler(item, context);
+                // if the handler doesn't fail and doesn't return false, continue
+                if (result !== false) {
+                  continue;
+                }
+              } catch {}
+              // if the handler fails or returns false, return the rest of the items
+              return itemGroup.slice(Number(i)).map((i) => i.id);
```

Comment:
Is the id different for each item even when the key is the same?

Reply:
yes, or I assume so, they are unique items

keys: [event1, event2]

if event 1 fails, so should 2
if event 2 fails, 1 isn't repeated.